transport.go

```diff
@@ -150,13 +150,31 @@ func (t *transport) handle() {
 	for {
 		select {
 		case req := <-t.requests:
-			// BUG(stevvooe): This is an awful tag allocation procedure.
-			// Replace this with something that let's us allocate tags and
-			// associate data with them, returning to them to a pool when
-			// complete. Such a system would provide a lot of information
-			// about outstanding requests.
-			tags++
-			fcall := newFcall(tags, req.message)
+			var selected Tag
+			if req.message.Type() == Tversion {
+				// Tversion can only use NOTAG.
+				if _, exists := outstanding[NOTAG]; exists {
+					req.err <- errors.New("NOTAG already in use")
+				}
+				selected = NOTAG
+			} else {
+				// The tag pool is depleted if all 65536 tags are taken, or if
+				// 65535 tags are taken and NOTAG is available.
+				if len(outstanding) > 0xFFFF {
+					req.err <- errors.New("tag pool depleted")
+				} else if len(outstanding) == 0xFFFF {
+					if _, exists := outstanding[NOTAG]; !exists {
+						req.err <- errors.New("tag pool depleted")
+					}
+				}
+
+				exists := true
+				for selected = 0; exists && selected < NOTAG; selected++ {
```

Comment:
I know what a slice is. I said array intentionally. :)

In the simple mode, the slice cap would be equal to or greater than the highest tag in use. The slice cannot shrink without manually replacing it with a new slice with copied content, whereas the map can do whatever it wants (Not sure what it does in this scenario, actually). Furthermore, the slice will over-grow (512KB + 25% worst-case, IIRC) if we append.

We could make a `tagpool` type to take care of all this logic (check for slot, grow if none found and len < NOTAG, search from end to first tag to slice and shrink if len is much shorter than cap and cap is >100), but... That's a benchmark and an optimisation for another day. I prefer to have easily reproducible numbers when making performance claims and modifications.